### PR TITLE
Exercise 3

### PR DESCRIPTION
**Pull request by vaibhav1sh**
Pull request for Module 3 exercise.

### CODE REVIEW

File: tests_e2e/test_e2e.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    button_to_click = driver.find_element_by_xpath(
+        "/html/body/div/div[3]/div/form/ul/li[3]/button"
+        )
+    button_to_click.click()
+    assert (driver.page_source.find("Creating new card")) 
+
+def test_mark_card_complete(driver, test_app):
+    driver.get('http://localhost:5000/')
+    button_to_click = driver.find_element_by_xpath(
+        "/html/body/div/div[2]/div/ul[1]/li[1]/form/button"
```
Comment (JackMeadDev) [inline tests_e2e/test_e2e.py]: This feels a very flaky way to find your button - if anyone adjusts the page structure at all you'll be constantly updating this! Finding it by e.g. name/id might be a cleaner & more maintainable solution for this

File: tests/test_trello_api_calls.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+'''Unit Tests for Trello API - Create Board'''
+
+import pytest, os
+from todo_app.trello_api_calls import create_trello_board, delete_trello_board
+
+def test_create_delete_trello_board():
```
Comment (JackMeadDev) [inline tests/test_trello_api_calls.py:10]: The file describes this as a unit test, but it looks like you are expecting this to actually talk to Trello - i.e. it's more like an integration test. Importantly, that means this it requires having the real API key/token - so that should be an explicit part of this test set up (i.e. as with the end-to-end test, we should load the `.env` file)

File: tests/test_viewmodel.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+# Testing decorated version of get_cards
+@pytest.mark.parametrize('from_threshold, result',[
+    (midnight, 2),
+    (midnight + timedelta(hours=5), 1),
+    (midnight - timedelta(hours=7), 3)
+    ])
+def test_get_cards_from_threshold(prep_view_model, from_threshold, result):
+    assert len(prep_view_model.get_cards("Done", from_threshold)) == result
+
+@pytest.mark.parametrize('to_threshold, result',[
```
Comment (JackMeadDev) [inline tests/test_viewmodel.py:51]: Nice use of parameterised tests

File: tests_e2e/test_e2e.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+from selenium import webdriver
+from selenium.webdriver.support.ui import Select
+from threading import Thread
+from todo_app.trello_api_calls import create_trello_board, delete_trello_board,\
+    fetch_all_cards
+from todo_app.app import create_app
+
+@pytest.fixture(scope='module')
+def test_app():
+    result = create_trello_board()
```
Comment (JackMeadDev) [inline tests_e2e/test_e2e.py:15]: You aren't explicitly loading any environment variables in this file, which suggests that the test expects those to already exist. VSCode's test runner can inject these automatically which might be hiding the problem, but if you run the tests from the command line (with `poetry run pytest`) then the tests will fail

We can fix this by loading `.env` during this setup (before creating the trello board) in a similar fashion to how you loaded `.env.test` in the integration test

File: todo_app/trello_api_calls.py [hunk 5]
```diff
@@ -58,4 +66,25 @@ def change_card_status(card_id,to_state="Done"):
...
     listdata = {"idList" : list_id}
-    api_response = requests.put(url,params=payload, data = listdata)
+    requests.put(url,params=payload, data = listdata)
+
+# Written solely for the purpose of testing. 
+def create_trello_board():
+    payload = {
+        'key': os.environ.get('TRELLO_KEY'),
+        'token': os.environ.get('TRELLO_TOKEN'),
+        'name': os.environ.get('TRELLO_BOARD_NAME_CREATE')
```
Comment (JackMeadDev) [inline todo_app/trello_api_calls.py:76]: This environment variable doesn't seem to be documented anywhere - we should add it to the README/`.env.template` files so that future devs would know this is required for the tests

File: todo_app/view_model.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
+from datetime import datetime, timedelta, time
+
+# Decorator for get_cards function to slice result by time.
+def split_result_by_updated_time(func):
```
Comment (JackMeadDev) [inline todo_app/view_model.py:4]: This is a really interesting way of solving this issue, but it looks pretty complex!

Note that a slightly different choice of default argument could make this much simpler - by switching from None defaults to `datetime.min` & `datetime.max` respectively, we can always assume we use the last of the checks
```python
def split_result_by_updated_time(func):
    def wrapper_split_result_by_updated_time(self, list_category, \
        from_threshold = datetime.min, to_threshold = datetime.max):
        intermediate_list = func(self,list_category)
        final_list = []
        for x in intermediate_list:
            if from_threshold <= x.updated_time < to_threshold: 
                final_list.append(x)
        return final_list
    return wrapper_split_result_by_updated_time
```

File: todo_app/view_model.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        result_cards = []
+        if list_category == "All":
+            return self._cards
+        else:
+            for x in self._cards:
+                if x.status == list_category:
+                    result_cards.append(x)
+            return result_cards
+
+    def return_all_done_cards(self, limit_of_cards):
```
Comment (JackMeadDev) [inline todo_app/view_model.py:35]: Ideally, we'd separate out some of this complexity, by having all of the return values as separate properties, rather than combining them all in one list

e.g. one property for `all_done_cards`, another for `show_all_done_cards` etc.

File: .gitignore [hunk 8]
```diff
@@ -46,6 +46,9 @@ MANIFEST
 pip-log.txt
 pip-delete-this-directory.txt
 
+# Geckodriver log
+geckodriver.log
```
Comment (JackMeadDev) [inline .gitignore:50]: It would be good to also ignore the `geckodriver.exe` file here since that is largely & will depend on the users system. Also note that `geckodriver.log` was already checked in before you added this & this hasn't removed the file. To get rid of that, you can run `git rm --cached geckodriver.log`

**Review by JackMeadDev — CHANGES_REQUESTED** [review]
Nice work, your tests look thorough and break down the different levels well. I've left some thoughts on the review below, but the most important things would be:
* Our README should include instructions on how to run the tests, and mention any dependencies (e.g. firefox)
* The e2e test (and test_trello_api_calls) should pass if run directly in the command line (e.g. with `poetry run pytest`) - see thoughts below for that but it looks like you're not actively loading in your `.env` file

File: tests_e2e/test_e2e.py (hunk 1 above)
Comment (vaibhav1sh) [reply tests_e2e/test_e2e.py]: Agreed, replaced this with find_element_by_id

File: tests/test_trello_api_calls.py (hunk 2 above)
Comment (vaibhav1sh) [reply tests/test_trello_api_calls.py:10]: Added

File: tests_e2e/test_e2e.py (hunk 4 above)
Comment (vaibhav1sh) [reply tests_e2e/test_e2e.py:15]: This caught me unaware - the tests were failing through command line ... Fixed !

File: todo_app/view_model.py (hunk 6 above)
Comment (vaibhav1sh) [reply todo_app/view_model.py:4]: datetime.min and datetime.max were good replacements for None, and I was able to eliminate two checks 👍 .. However, I retained one of the conditions as a check to ensure function doesn't unnecessarily iterate over all the cards/items when it doesn't need to. Thanks !

File: tests/test_viewmodel.py (hunk 3 above)
Comment (vaibhav1sh) [reply tests/test_viewmodel.py:51]: Woohoo! 🥇

File: todo_app/trello_api_calls.py (hunk 5 above)
Comment (vaibhav1sh) [reply todo_app/trello_api_calls.py:76]: Agreed, added to `.env.template` and `.env.test`

File: .gitignore (hunk 8 above)
Comment (vaibhav1sh) [reply .gitignore:50]: Done!

File: todo_app/view_model.py (hunk 7 above)
Comment (vaibhav1sh) [reply todo_app/view_model.py:35]: This one has me stumped. I read about descriptors and properties but I am unable to apply properties in this context. This is going to be one long-winded answer. 

 **TL;DR:**  What benefit do I get from accessing these return values as properties, that I do not get when I access these as return value of their individual functions (e.g., function show_all_done_cards() as in Step 1)

**Long Version**
- Specifically in the context of this file - I thought properties are only useful if the class attribute needs to have additional logic implemented before it is set (or read, etc.), or if an attribute isn't known at the time when object is instantiated. 
- I tested a few cases as per below -  

> **Step 1:** Let us just break down these return values by putting them in a function of their own. I simplified the file, and looked at just one value, i.e., `show_all_done_cards`
```
class ViewModel:
    def __init__(self, cards: list):
        self._cards = cards

    def get_cards(self):
        return self._cards
    
    def show_all_done_cards(self):
            all_done_cards = self.get_cards()
            if len(all_done_cards) > 5:
                show_all_done_cards = False
            else:
                show_all_done_cards = True
            return show_all_done_cards

viewmodel_obj = ViewModel(["Card 1", "Card 2", "Card 3"])
print("OPTION 1: ", viewmodel_obj.show_all_done_cards())
```
> **Step 2:** I then changed the function to property. As a result, nothing really changes except for the fact that I can now access `show_all_done_cards` as an attribute instead of a function in the print statement. No other difference that I can see ..  
```
    @property
    def show_all_done_cards(self):
            all_done_cards = self.get_cards()
            if len(all_done_cards) > 5:
                show_all_done_cards = False
            else:
                show_all_done_cards = True
            return show_all_done_cards

viewmodel_obj = ViewModel(["Card 1", "Card 2", "Card 3"])
print("OPTION 2: ",viewmodel_obj.show_all_done_cards)
```
> **Step 3:** Desperate to see the benefit of previous step, I added `show_all_done_cards` as attribute. I thought this may be more OOPS-like design and the value can be stored as attribute of the object. 

I then realized that below code does not work because I cannot **set** the value through the **getter** method of property object.

```
class ViewModel:
    def __init__(self, cards: list):
        self._cards = cards
        self._show_all_done_cards = None

    @property
    def show_all_done_cards(self):
            all_done_cards = self.get_cards()
            if len(all_done_cards) > 5:
                show_all_done_cards = False
            else:
                show_all_done_cards = True
            self._show_all_done_cards = show_all_done_cards # Fails here because it can't set the value 
            return self._show_all_done_cards
```

> **Step 4:** Next, I tried to do it the standard way. Write a property, and write a setter as well.  However, I still have questions, and I can't see the benefits of this approach - 

- The below code has a getter and setter. However, the setter has to work before the getter method, otherwise the getter method returns `None`. 
- There is no apparent benefit in storing `show_all_done_cards` as object attribute because the value of this flag may change anytime, hence has to be always calculated through the logic in setter method. Why to store and set this then? 

```
class ViewModel:
    def __init__(self, cards: list):
        self._cards = cards
        self._show_all_done_cards = None

    @property
    def show_all_done_cards(self):
            return self._show_all_done_cards

    @show_all_done_cards.setter
    def show_all_done_cards(self):
            all_done_cards = self.get_cards()
            if len(all_done_cards) > 5:
                show_all_done_cards = False
            else:
                show_all_done_cards = True
            self._show_all_done_cards = show_all_done_cards

viewmodel_obj = ViewModel(["Card 1", "Card 2", "Card 3"])
print("OPTION 4: ",viewmodel_obj.show_all_done_cards)  #This will print None
print(viewmodel_obj.__dict__) # verified object attributes
```

**Review by JackMeadDev — APPROVED** [review]
Thanks for making those changes, I can now run the tests without issue (including from a terminal) following your Readme - I'm happy for you to merge this when you're ready
